src/Microsoft.AspNet.OData.Shared/Query/ODataPathQueryBuilder.cs

```diff
@@ -0,0 +1,208 @@
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using Microsoft.OData.Edm;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.AspNet.OData.Query
+{
+
+    /// <summary>
+    /// This class transforms an <see cref="IQueryable"/> based on
+    /// the key and property accesses in the segments of an <see cref="ODataPath"/>
+    /// </summary>
+    internal class ODataPathQueryBuilder
+    {
+        private readonly IQueryable source;
+        private readonly Routing.ODataPath path;
+
+        /// <summary>
+        /// Creates an instance of <see cref="ODataPathQueryBuilder"/>
+        /// </summary>
+        /// <param name="source">The original <see cref="IQueryable"/> to be transformed</param>
+        /// <param name="model">The <see cref="IEdmModel"/></param>
+        /// <param name="path">The <see cref="ODataPath"/> based on which to transform the <see cref="source"/></param>
+        public ODataPathQueryBuilder(IQueryable source, IEdmModel model, Routing.ODataPath path)
+        {
+            this.source = source;
+            this.path = path;
+        }
+
+        /// <summary>
+        /// Transforms the source <see cref="IQueryable"/> based on the sequence of path segments in the <see cref="ODataPath"/>
+        /// </summary>
+        /// <returns>The result of the query transformtions or null if the path contained unsupported segments</returns>
+        public ODataPathQueryResult BuildQuery()
+        {
+            ODataPathQueryResult result = new ODataPathQueryResult();
+
+            IEnumerable<ODataPathSegment> segments = path.Segments;
+
+            IQueryable queryable = source;
+            
+            ODataPathSegment firstSegment = segments.FirstOrDefault();
+
+            if (!(firstSegment is EntitySetSegment || firstSegment is SingletonSegment))
+            {
+                return null;
+            }
+
+            IEnumerable<ODataPathSegment> remainingSegments = segments.Skip(1);
+
+            foreach (ODataPathSegment segment in remainingSegments)
+            {
+                if (segment is KeySegment keySegment)
+                {
+                    // filterPredicate = entity => (entity.KeyProp1 == Val1) && (entity.keyProp2 == Val2) && ...
+                    ParameterExpression filterParam = Expression.Parameter(queryable.ElementType, "entity");
+                    IEnumerable<BinaryExpression> conditions = keySegment.Keys.Select(kvp =>
+                        Expression.Equal(
+                            Expression.Property(filterParam, kvp.Key),
+                            Expression.Constant(kvp.Value)));
+                    BinaryExpression filterBody = conditions.Aggregate((left, right) => Expression.AndAlso(left, right));
+                    LambdaExpression filterPredicate = Expression.Lambda(filterBody, filterParam);
+
+                    // queryable = queryable.Where(entity => (entity.KeyProp1 == Val1) && (entity.KeyProp2 == Val2) ...)
```

Comment:
remove commented code

Reply:
This was not commented code, it was meant to explain what the following statement was trying to achieve using reflection:
```c#
queryable = ExpressionHelpers.Where(queryable, filterPredicate, queryable.ElementType);
```
maybe I should make that more clear.